libs/dkg/include/dkg/rbc_messages.hpp

```diff
@@ -94,163 +102,82 @@ class RBCMessage
 };
 
-class RBroadcast : public RMessage
+class RMessage : public RBCMessage
 {
 public:
-  RBroadcast(uint16_t channel, uint32_t id, uint8_t counter, SerialisedMessage msg)
-    : RMessage{channel, id, counter, std::move(msg), MessageType::RBROADCAST}
-  {}
-  explicit RBroadcast(RBCSerializer &serialiser)
-    : RMessage{MessageType::RBROADCAST}
-  {
-    serialiser >> channel_ >> id_ >> counter_ >> message_;
-  }
-  ~RBroadcast() override = default;
+  using RBCMessage::RBCMessage;
+  virtual ~RMessage() = default;
+  using RBCMessage::message;
 };
 
-class REcho : public RHash
+template <RBCMessage::MessageType TYPE>
+class RMessageImpl final : public RMessage
 {
 public:
-  REcho(uint16_t channel, uint32_t id, uint8_t counter, TruncatedHash msg_hash)
-    : RHash{channel, id, counter, std::move(msg_hash), MessageType::RECHO}
-  {}
-  explicit REcho(RBCSerializer &serialiser)
-    : RHash{MessageType::RECHO}
+  using RMessage::RMessage;
+  virtual ~RMessageImpl() = default;
+
+  MessageType type() const override
   {
-    serialiser >> channel_ >> id_ >> counter_ >> hash_;
+    return TYPE;
   }
-  ~REcho() override = default;
 };
 
-class RReady : public RHash
+class RHash : public RBCMessage
 {
 public:
-  // Construction/Destruction
-  RReady(uint16_t channel, uint32_t id, uint8_t counter, TruncatedHash msg_hash)
-    : RHash{channel, id, counter, std::move(msg_hash), MessageType::RREADY}
-  {}
-  explicit RReady(RBCSerializer &serialiser)
-    : RHash{MessageType::RREADY}
-  {
-    serialiser >> channel_ >> id_ >> counter_ >> hash_;
-  }
-  ~RReady() override = default;
+  // Destruction
+  using RBCMessage::hash;
+  using RBCMessage::RBCMessage;
+  virtual ~RHash() = default;
 };
 
-class RRequest : public RBCMessage
+template <RBCMessage::MessageType TYPE>
+class RHashImpl final : public RHash
 {
 public:
-  RRequest(uint16_t channel, uint32_t id, uint8_t counter)
-    : RBCMessage{channel, id, counter, MessageType::RREQUEST}
-  {}
-  explicit RRequest(RBCSerializer &serialiser)
-    : RBCMessage{MessageType::RREQUEST}
-  {
-    serialiser >> channel_ >> id_ >> counter_;
-  }
-  ~RRequest() override = default;
+  using RHash::RHash;
+  virtual ~RHashImpl() = default;
 
-  RBCSerializer Serialize() const override
+  MessageType type() const override
   {
-    RBCSerializer serialiser;
-    serialiser << channel_ << id_ << counter_;
-    return serialiser;
+    return TYPE;
   }
 };
 
-class RAnswer : public RMessage
-{
-public:
-  RAnswer(uint16_t channel, uint32_t id, uint8_t counter, SerialisedMessage msg)
-    : RMessage{channel, id, counter, std::move(msg), MessageType::RANSWER}
-  {}
-  explicit RAnswer(RBCSerializer &serialiser)
-    : RMessage{MessageType::RANSWER}
-  {
-    serialiser >> channel_ >> id_ >> counter_ >> message_;
-  }
-  ~RAnswer() override = default;
-};
+using RBroadcast = RMessageImpl<RBCMessage::MessageType::R_BROADCAST>;
+using RRequest   = RMessageImpl<RBCMessage::MessageType::R_REQUEST>;
```

Comment:
A request does not contain any message or hash. It is only needs to contain the channel, id and counter.

Reply:
It is worth adding an empty payload to request as this streamlines all message forms and will simplify the design significantly in coming PRs. The overhead of doing this is 4 bytes.